src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.cpp

```diff
@@ -0,0 +1,893 @@
+  messaget log(message_handler);
+  if(loop_contract_mode != dfcc_loop_contract_modet::NONE)
+  {
+    dfcc_check_loop_normal_form(goto_program, log);
+    loop_nesting_graph = build_loop_nesting_graph(goto_program, log);
+    check_inner_loop_have_contracts(loop_nesting_graph);
+
+    auto topsorted = loop_nesting_graph.topsort();
+    for(const auto idx : topsorted)
+    {
+      topsorted_loops.push_back(idx);
+    }
+  }
+
+  tag_loop_instructions(goto_program, loop_nesting_graph);
+
+  // initialize attributes using auxiliary class
+  gen_dfcc_loop_infot(
+    function_id,
+    goto_function,
+    goto_model,
+    utils,
+    library,
+    message_handler,
+    loop_nesting_graph,
+    loop_info_map,
+    ident_to_loop_id_map,
+    top_level_loops,
+    top_level_local,
+    top_level_tracked);
+}
+
+void dfcc_cfg_infot::output(
+  std::ostream &out,
+  const namespacet &ns,
+  const irep_idt &mode) const
+{
+  out << "// dfcc_cfg_infot for:  " << function_id << "\n";
+  out << "// top_level_local: {";
+  for(auto &id : top_level_local)
+  {
+    out << id << ", ";
+  }
+  out << "}\n";
+
+  out << "// top_level_tracked: {";
+  for(auto &id : top_level_tracked)
+  {
+    out << id << ", ";
+  }
+  out << "}\n";
+
+  out << "// loop:\n";
+  for(auto &loop : loop_info_map)
+  {
+    out << "// dfcc-loop_id:" << loop.first << "\n";
+    auto head = loop.second.find_head(goto_function.body);
+    auto latch = loop.second.find_latch(goto_function.body);
+    out << "// head:\n";
+    head.value()->output(out);
+    out << "// latch:\n";
+    latch.value()->output(out);
+    loop.second.output(out, ns, mode);
+  }
+  out << "// program:\n";
+  Forall_goto_program_instructions(target, goto_function.body)
+  {
+    out << "// dfcc-loop-id:" << dfcc_get_loop_id(target).value();
+    out << " cbmc-loop-number:" << target->loop_number;
+    out << " top-level:" << dfcc_is_loop_top_level(target);
+    out << " head:" << dfcc_is_loop_head(target);
+    out << " body:" << dfcc_is_loop_body(target);
+    out << " exiting:" << dfcc_is_loop_exiting(target);
+    out << " latch:" << dfcc_is_loop_latch(target);
+    out << "\n";
+    target->output(out);
+  }
+}
+
+const exprt &
+dfcc_cfg_infot::get_write_set(goto_programt::const_targett target) const
+{
+  auto loop_id_opt = dfcc_get_loop_id(target);
+  PRECONDITION(
+    loop_id_opt.has_value() &&
+    is_valid_loop_or_top_level_id(loop_id_opt.value()));
+  auto loop_id = loop_id_opt.value();
+  if(is_top_level_id(loop_id))
+  {
+    return top_level_write_set;
+  }
+  else
+  {
+    return loop_info_map.at(loop_id).addr_of_write_set_var;
+  }
+}
+
+const symbol_exprt &
+dfcc_cfg_infot::get_loop_write_set(goto_programt::const_targett target) const
+{
+  auto loop_id_opt = dfcc_get_loop_id(target);
+  PRECONDITION(
+    loop_id_opt.has_value() && is_valid_loop_id(loop_id_opt.value()));
+  auto loop_id = loop_id_opt.value();
+  return loop_info_map.at(loop_id).addr_of_write_set_var;
+}
```

Comment:
See above: I think there should only be one such function here, and all the which-loop-info-map-object-is-it should be hidden behind an interface.

Reply:
I'm adding three new methods to hide the disjunctions.